src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -267,34 +305,175 @@ private ValueTask<(HttpConnection, HttpResponseMessage)> GetConnectionAsync(Http
             }
         }
 
+        private async ValueTask<(HttpConnectionBase connection, bool isNewConnection, HttpResponseMessage failureResponse)>
+            GetHttp2ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Debug.Assert(_kind == HttpConnectionKind.Https || _kind == HttpConnectionKind.SslProxyTunnel);
+
+            // See if we have an HTTP2 connection
+            lock (SyncObj)
+            {
+                if (_http2Connection != null)
+                {
+                    // TODO: Check connection validity and timeouts
+
+                    if (NetEventSource.IsEnabled)
+                    {
+                        Trace("Using existing HTTP2 connection.");
+                    }
+
+                    return (_http2Connection, false, null);
+                }
+
+                // Ensure that the connection creation semaphore is created 
+                if (_http2ConnectionCreateLock == null)
+                {
+                    _http2ConnectionCreateLock = new SemaphoreSlim(1);
+                }
+            }
+
+            // Try to establish an HTTP2 connection
+            Socket socket = null;
+            SslStream sslStream = null;
+            TransportContext transportContext = null;
+
+            // Serialize creation attempt
+            await _http2ConnectionCreateLock.WaitAsync().ConfigureAwait(false);
```

Comment:
Why do we need both SyncObj and _http2ConnectionCreateLock?  Should they be one in the same?